# **HCDE 530 A5 - Pandas Analysis**

## **Background**

#### **My Dataset**: 
I'll be using the Jikan API (https://jikan.moe/), an unofficial REST API that pulls data from MyAnimeList.

#### **Analytical Questions:**
1. Among the 10 studios with the most titles in the top 100 anime, which has the highest average user score?
2. Among the top 100 anime, which 5 genres appear most frequently?
3. Among the top 100 anime, which decade of release (e.g., 2000s, 2010s, 2020s) has the highest average user score?

---
## **Part A: Setup and Load Data**

Run this cell first. It imports pandas and confirms your version.


In [11]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

print("pandas version:", pd.__version__)

pandas version: 3.0.2


In [12]:
df = pd.read_csv('top_anime_100.csv')
df.head(5)

,mal_id,url,approved,titles,title,title_english,title_japanese,title_synonyms,type,source,...,aired.prop.from.month,aired.prop.from.year,aired.prop.to.day,aired.prop.to.month,aired.prop.to.year,aired.string,broadcast.day,broadcast.time,broadcast.timezone,broadcast.string
0,52991,https://myanimelist.net/anime/52991/Sousou_no_...,True,"[{'type': 'Default', 'title': 'Sousou no Frier...",Sousou no Frieren,Frieren: Beyond Journey's End,葬送のフリーレン,"['Frieren at the Funeral', 'Frieren The Slayer']",TV,Manga,...,9,2023,22.0,3.0,2024.0,"Sep 29, 2023 to Mar 22, 2024",Fridays,23:00,Asia/Tokyo,Fridays at 23:00 (JST)
1,61469,https://myanimelist.net/anime/61469/Steel_Ball...,True,"[{'type': 'Default', 'title': 'Steel Ball Run:...",Steel Ball Run: JoJo no Kimyou na Bouken,Steel Ball Run: JoJo's Bizarre Adventure,スティール・ボール・ラン ジョジョの奇妙な冒険,"[""JoJo's Bizarre Adventure Part 7: Steel Ball ...",ONA,Manga,...,3,2026,NaN,NaN,NaN,"Mar 19, 2026 to ?",NaN,NaN,NaN,NaN
2,5114,https://myanimelist.net/anime/5114/Fullmetal_A...,True,"[{'type': 'Default', 'title': 'Fullmetal Alche...",Fullmetal Alchemist: Brotherhood,Fullmetal Alchemist: Brotherhood,鋼の錬金術師 FULLMETAL ALCHEMIST,['Hagane no Renkinjutsushi: Fullmetal Alchemis...,TV,Manga,...,4,2009,4.0,7.0,2010.0,"Apr 5, 2009 to Jul 4, 2010",Sundays,17:00,Asia/Tokyo,Sundays at 17:00 (JST)
3,57555,https://myanimelist.net/anime/57555/Chainsaw_M...,True,"[{'type': 'Default', 'title': 'Chainsaw Man Mo...",Chainsaw Man Movie: Reze-hen,Chainsaw Man – The Movie: Reze Arc,劇場版 チェンソーマン レゼ篇,['Gekijouban Chainsaw Man: Reze-hen'],Movie,Manga,...,9,2025,NaN,NaN,NaN,"Sep 19, 2025",NaN,NaN,NaN,NaN
4,9253,https://myanimelist.net/anime/9253/Steins_Gate,True,"[{'type': 'Default', 'title': 'Steins;Gate'}, ...",Steins;Gate,Steins;Gate,STEINS;GATE,[],TV,Visual novel,...,4,2011,14.0,9.0,2011.0,"Apr 6, 2011 to Sep 14, 2011",Wednesdays,02:05,Asia/Tokyo,Wednesdays at 02:05 (JST)


---

## **Part B: Answer Analytical Questions**

### **#1. Among the 10 studios with the most titles in the top 100 anime, which has the highest average user score?**

In [ ]:
# Retrieve the top 10 studios with the highest average user score,
# then sort them from highest to lowest score
df.groupby('studios')['score'].mean().sort_values(ascending=False).head(10)

studios
[{'mal_id': 287, 'type': 'anime', 'name': 'David Production', 'url': 'https://myanimelist.net/anime/producer/287/David_Production'}]                                                                                                                                              9.140000
[{'mal_id': 1269, 'type': 'anime', 'name': 'K-Factory', 'url': 'https://myanimelist.net/anime/producer/1269/K-Factory'}, {'mal_id': 2256, 'type': 'anime', 'name': 'Kitty Film Mitaka Studio', 'url': 'https://myanimelist.net/anime/producer/2256/Kitty_Film_Mitaka_Studio'}]    9.020000
[{'mal_id': 314, 'type': 'anime', 'name': 'White Fox', 'url': 'https://myanimelist.net/anime/producer/314/White_Fox'}]                                                                                                                                                            8.995000
[{'mal_id': 1258, 'type': 'anime', 'name': 'Bandai Namco Pictures', 'url': 'https://myanimelist.net/anime/producer/1258/Bandai_Namco_Pictures'}

In [ ]:
# Reorganize the data above into a tidy table for readability, while omitting mal_id, 
# type, anime, name, and URL as these details are irrelevant to the question

import ast
import pandas as pd
from IPython.display import display

def studio_names(cell):
    if pd.isna(cell):
        return []
    try:
        lst = ast.literal_eval(cell)
    except (ValueError, SyntaxError):
        return []
    return [s["name"] for s in lst]

tmp = df.assign(studio_name=df["studios"].apply(studio_names)).explode("studio_name")

top10_by_score = (
    tmp.groupby("studio_name", as_index=False)
    .agg(mean_score=("score", "mean"), title_count=("mal_id", "count"))
    .sort_values("mean_score", ascending=False)
    .head(10)
    .reset_index(drop=True)
)

display(top10_by_score)

,studio_name,mean_score,title_count
0,David Production,9.140000,1
1,K-Factory,9.020000,1
2,Kitty Film Mitaka Studio,9.020000,1
3,White Fox,8.995000,2
4,Bandai Namco Pictures,8.954000,5
5,TMS Entertainment,8.930000,1
6,Madhouse,8.894286,7
7,Sunrise,8.892857,7
8,TOHO animation STUDIO,8.880000,2
9,Bones,8.870000,3


**Note:**
After looking at the data in the above table, I realized that the suggested code was not correctly answering my question. My question asked, “Among the top 10 studios with the most titles...,” but it seems the Cursor agent only sorted studios by highest average user score rather than accounting for the number of titles produced by each studio.

In [ ]:
# First find studios with the most titles, then compare average user score

import ast
import pandas as pd
from IPython.display import display

# Give me a simple list of studio names
def studio_names(cell):
    if pd.isna(cell):
        return []
    try:
        lst = ast.literal_eval(cell)
    except (ValueError, SyntaxError):
        return []
    return [s["name"] for s in lst]

# Turn that list into 1 row per studio, so I can count and average by studio
tmp = df.assign(studio_name=df["studios"].apply(studio_names)).explode("studio_name")

# Summarize every studio with average score and number of titles
studio_stats = (
    tmp.groupby("studio_name", as_index=False)
    .agg(mean_score=("score", "mean"), title_count=("mal_id", "count"))
)

# Collect the top 10 studios with the most titles and their average scores
top10_most_titles = (
    studio_stats.sort_values(["title_count", "mean_score"], ascending=[False, False])
    .head(10)
    .reset_index(drop=True)
)

display(top10_most_titles)

# Show the studio that has the highest average score, which answers my question
best_avg_in_top10 = top10_most_titles.sort_values("mean_score", ascending=False).head(1)
display(best_avg_in_top10)

,studio_name,mean_score,title_count
0,MAPPA,8.811250,8
1,Madhouse,8.894286,7
2,Sunrise,8.892857,7
3,Studio Pierrot,8.761429,7
4,Bandai Namco Pictures,8.954000,5
5,Kyoto Animation,8.822000,5
6,Shaft,8.790000,5
7,A-1 Pictures,8.760000,4
8,Studio Signpost,8.742500,4
9,Bones,8.870000,3


,studio_name,mean_score,title_count
4,Bandai Namco Pictures,8.954,5


##### **Answer: Out of the top 10 studios with the most titles in the top 100 anime, Bandai Namco Pictures has the highest average score of 8.954, with 5 titles represented.**

### **#2. Among the top 100 anime, which 5 genres appear most frequently?**

In [ ]:
import ast
from IPython.display import display

# Produce a clean list of genre names
def genre_names(cell):
    if pd.isna(cell):
        return []
    try:
        lst = ast.literal_eval(cell)
    except (ValueError, SyntaxError):
        return []
    return [g["name"] for g in lst]

# Tell me which 5 genres occur most often and how many times each occur
top5_genre_names = (
    df.assign(genre_name=df["genres"].apply(genre_names))
    .explode("genre_name")
    ["genre_name"]
    .value_counts()
    .head(5)
    .rename_axis("genre")
    .reset_index(name="count")
)

display(top5_genre_names)

,genre,count
0,Drama,47
1,Action,44
2,Adventure,22
3,Supernatural,19
4,Comedy,19


##### **Answer: The 5 genres that appear most frequently in the top 100 anime are: Drama (47), Action (44), Adventure (22), Supernatural (19), and Comedy (19).**

### **#3: Among the top 100 anime, which decade of release (e.g., 2000s, 2010s, 2020s) has the highest average user score?**

In [ ]:
import pandas as pd
from IPython.display import display

# Keep anime with a known year in 1980-2026
period = df.dropna(subset=["year"]).loc[lambda d: (d["year"] >= 1980) & (d["year"] <= 2026)].copy()

# Bucket anime into 5 decade ranges
period["decade"] = pd.cut(
    period["year"],
    bins=[1979, 1989, 1999, 2009, 2019, 2026],
    labels=["1980-1989", "1990-1999", "2000-2009", "2010-2019", "2020-2026"],
    include_lowest=True,
    right=True,
)

# Compute average score and count per range, and sort by average score
score_by_decade = (
    period.groupby("decade", observed=True)
    .agg(mean_score=("score", "mean"), anime_count=("mal_id", "count"))
    .reset_index()
    .sort_values("mean_score", ascending=False)
)

display(score_by_decade)

# Report the top decade
winner = score_by_decade.iloc[0]
print(
    f"Highest average score: {winner['decade']} "
    f"(mean score {winner['mean_score']:.4f}, n={int(winner['anime_count'])})"
)

,decade,mean_score,anime_count
2,2000-2009,8.823000,10
3,2010-2019,8.820435,23
0,1980-1989,8.820000,1
4,2020-2026,8.782500,32
1,1990-1999,8.720000,3


Highest average score: 2000-2009 (mean score 8.8230, n=10)


**Note:** When I reviewed the table and added the counts, I realized they only summed to 69, even though there should have been a total of 100 anime. I then asked Cursor where the other 31 entries went, and it revealed that those anime did not have a year listed.

In [ ]:
# Missing year -> "Unknown". Still excludes non-null years before 1980 or after 2026.

# Check through titles to see if they have a listed year between 1980 and 2026
has_year = df["year"].notna()
in_range = (df["year"] >= 1980) & (df["year"] <= 2026)

# For animes with known years, categorize them into decade range
bucketed = df.loc[has_year & in_range].copy()
bucketed["decade"] = pd.cut(
    bucketed["year"],
    bins=[1979, 1989, 1999, 2009, 2019, 2026],
    labels=["1980-1989", "1990-1999", "2000-2009", "2010-2019", "2020-2026"],
    include_lowest=True,
    right=True,
)

# Take anime with unlisted year and put them together in Unknown group
unknown_year = df.loc[~has_year].copy()
unknown_year["decade"] = "Unknown"

period_with_unknown = pd.concat([bucketed, unknown_year], ignore_index=True)

# Compute average score and count per range, and sort by average score
score_by_decade_with_unknown = (
    period_with_unknown.groupby("decade", observed=True)
    .agg(mean_score=("score", "mean"), anime_count=("mal_id", "count"))
    .reset_index()
    .sort_values("mean_score", ascending=False)
)

display(score_by_decade_with_unknown)

# Display decade with top average score
winner_u = score_by_decade_with_unknown.iloc[0]
print(
    f"Highest average score: {winner_u['decade']} "
    f"(mean score {winner_u['mean_score']:.4f}, n={int(winner_u['anime_count'])})"
)

,decade,mean_score,anime_count
2,2000-2009,8.823000,10
3,2010-2019,8.820435,23
0,1980-1989,8.820000,1
5,Unknown,8.784839,31
4,2020-2025,8.782500,32
1,1990-1999,8.720000,3


Highest average score: 2000-2009 (mean score 8.8230, n=10)


##### **Answer: The decade with the highest average score is 2000–2009, with an average score of 8.8230 across 10 anime titles.**